### PR TITLE
IMPRV/Sorting-API

### PR DESCRIPTION
**Pull request by Dwaynekj**
@joneit @rdepena Paired PR with https://github.com/openfin/fin-hypergrid/pull/342/files

### CODE REVIEW

File: js/Base.js [hunk 1]
```diff
@@ -107,6 +107,14 @@ Base.prototype = {
             return this.dataSource.isDrillDown.apply(this.dataSource, arguments);
         }
     },
+    getReIndexedData: function() {
```
Comment (joneit) [inline js/Base.js]: We're moving this to MySorter

File: js/DataSourceTreeviewSorter.js [hunk 2]
```diff
@@ -20,15 +20,17 @@ var DataSourceTreeviewSorter = DataSourceSorterComposite.extend('DataSourceTreev
         var joined = this.treeview.viewMakesSense(),
             each = this.dataSource,
             last, // last sort spec ("first" sort) when and only when joined AND it is the group column
-            lastIsGroup, columnIndex, direction;
+            lastIsGroup, columnIndex, direction,
+             //This line is a HACK to allow sorting to work in Hypergrid while also ensuring the existing api/test suite is enabled
+            sorts = (this.sorter && this.sorter.state && this.sorter.state.sorts) || this.sorts;
```
Comment (joneit) [inline js/DataSourceTreeviewSorter.js]: see above: should be `this.sorts`

File: js/Base.js (hunk 1 above)
Comment (Dwaynekj) [inline js/Base.js]: DONE

File: js/DataSourceTreeviewSorter.js (hunk 2 above)
Comment (Dwaynekj) [inline js/DataSourceTreeviewSorter.js]: DONE

File: js/util/stableSort.js [hunk 3]
```diff
@@ -120,11 +120,13 @@ function sort(index, getValue, direction) {
             case undefined: // eslint-disable-line no-fallthrough
                 direction = 1;
             case 1:  // eslint-disable-line no-fallthrough
-                compare = ascending(typeof getValue(0));
+                // compare = ascending(typeof getValue(0));
+                compare = ascending(type);
```
Comment (Dwaynekj) [inline js/util/stableSort.js]: must always pass the type now...

File: js/DataSourceSorter.js [hunk 4]
```diff
@@ -13,10 +13,10 @@ var DataSourceSorter = DataSourceIndexed.extend('DataSourceSorter', {
      * @param {number} columnIndex
      * @param {number} [direction=1]
      */
-    sortOn: function(columnIndex, direction) {
+    sortOn: function(columnIndex, direction, type) {
         var dataSource = this.dataSource,
             columnName = dataSource.getFields()[columnIndex],
```
Comment (joneit) [inline js/DataSourceSorter.js]: We should not be using `getFields` anymore unless we really do need an array of field names. Use `dataSource.schema[columnIndex].name` instead.

File: js/DataSourceSorter.js [hunk 5]
```diff
@@ -13,10 +13,10 @@ var DataSourceSorter = DataSourceIndexed.extend('DataSourceSorter', {
...
-            calculator = dataSource.getProperty('calculators')[columnIndex];
+            calculator = dataSource.schema[columnIndex] && dataSource.schema[columnIndex].calculator;
```
Comment (joneit) [inline js/DataSourceSorter.js]: There is no reason to check for schema element. It should always be defined.

File: js/util/stableSort.js [hunk 6]
```diff
@@ -103,7 +103,7 @@ function descending(typeOfData) {
  * @param {function} getValue
  * @param {number} [direction=1]
  */
-function sort(index, getValue, direction) {
+function sort(index, getValue, direction, type) {
```
Comment (joneit) [inline js/util/stableSort.js]: We still need the ability to specify `comparitor.asc` and `comparitor.desc` per schema spec outlined in [#342](https://github.com/openfin/fin-hypergrid/pull/342).

File: js/DataSourceSorterComposite.js [hunk 7]
```diff
@@ -49,22 +49,42 @@ var DataSourceSorterComposite = DataSourceIndexed.extend('DataSourceSorterCompos
...
-        });
-
-        this.last = each;
+        // get list of sorts from either API or use existing
+        this.sorts = (this.sorter && this.sorter.prop('sorts')) || this.sorts;
+
+        if (this.sorts) {
+            this.sorts.forEach(function(sortSpec) {
+                each = new DataSourceSorter(each);
+                each.sortOn(sortSpec.columnIndex, sortSpec.direction, sortSpec.type);
```
Comment (joneit) [inline js/DataSourceSorterComposite.js]: For type we should be using `schema[sortSpec.columnIndex].type`.

Passing `type` in the `sortSpec` is OK I guess but shouldn't be necessary. So if it is passed, we need to decide which gets priority. I'm thinking the passed value would override the schema value.

File: js/util/stableSort.js (hunk 6 above)
Comment (Dwaynekj) [inline js/util/stableSort.js]: Wont be doing as discussed

File: js/DataSourceSorter.js (hunk 5 above)
Comment (Dwaynekj) [inline js/DataSourceSorter.js]: Done

File: js/DataSourceSorter.js (hunk 4 above)
Comment (Dwaynekj) [inline js/DataSourceSorter.js]: DONE

File: js/DataSourceSorterComposite.js (hunk 7 above)
Comment (Dwaynekj) [inline js/DataSourceSorterComposite.js]: Its coming from schema [here](https://github.com/openfin/fin-hypergrid/pull/358/files#diff-71f67d01316a34b59304606398d01185R56) in the add-on

**Pull request merged**